nada_dsl/nada_types/scalar_types.py

```diff
@@ -0,0 +1,458 @@
+    else:
+        operation = PublicOutputEquality(left=left, right=right, source_ref=SourceRef.back_frame())
+        return PublicBoolean(inner=operation)
+
+
+class BooleanType(ScalarType):
+    """This abstraction represents all boolean types:
+        - Boolean, PublicBoolean, SecretBoolean
+        It is used to overload the common operation for all specific types are mentioned above.
+     """
+
+    def __and__(self, other):
+        return binary_logical_operation("BooleanAnd", "&", self, other, lambda lhs, rhs: lhs & rhs)
+
+    def __or__(self, other):
+        return binary_logical_operation("BooleanOr", "|", self, other, lambda lhs, rhs: lhs | rhs)
+
+    def __xor__(self, other):
+        return binary_logical_operation("BooleanXor", "^", self, other, lambda lhs, rhs: lhs ^ rhs)
+
+    def if_else(self, arg_0: ScalarType, arg_1: ScalarType) -> ScalarType:
+        """This function implements the function 'if_else' for every class that extends 'BooleanType'."""
+        base_type = arg_0.to_base_type()
+        if base_type != arg_1.to_base_type() or base_type == BaseType.BOOLEAN or self.to_mode() == Mode.CONSTANT:
+            raise TypeError(f"Invalid operation: {self}.IfElse({arg_0}, {arg_1})")
+        mode = Mode(max([self.to_mode().value, arg_0.to_mode().value, arg_1.to_mode().value]))
+        operation = IfElse(this=self, arg_0=arg_0, arg_1=arg_1, source_ref=SourceRef.back_frame())
+        if mode == Mode.CONSTANT or mode == Mode.PUBLIC:
+            return new_public(base_type, operation)
+        else:
+            return new_secret(base_type, operation)
+
+
+def binary_logical_operation(operation, operator, left: ScalarType, right: ScalarType, f) -> ScalarType:
+    """This function is an abstraction for the logical operations."""
+    base_type = left.to_base_type()
+    if base_type != right.to_base_type() or not base_type == BaseType.BOOLEAN:
+        raise TypeError(f"Invalid operation: {left} {operator} {right}")
+    mode = Mode(max([left.to_mode().value, right.to_mode().value]))
+    if mode == Mode.CONSTANT:
+        return Boolean(value=bool(f(left.value, right.value)))
+    elif mode == Mode.PUBLIC:
+        operation = globals()[operation](left=left, right=right, source_ref=SourceRef.back_frame())
+        return PublicBoolean(inner=operation)
+    else:
+        operation = globals()[operation](left=left, right=right, source_ref=SourceRef.back_frame())
+        return SecretBoolean(inner=operation)
+
+
+@dataclass
+class Integer(NumericType):
+    value: int
+
+    def __init__(self, value: int):
+        super().__init__(inner=Literal(value=value, source_ref=SourceRef.back_frame()))
+        if isinstance(value, int):
+            self.value = value
+        else:
+            raise ValueError(f"Expected int, got {type(value).__name__}")
+
+    def to_base_type(self) -> BaseType:
+        return BaseType.INTEGER
+
+    def to_mode(self) -> Mode:
+        return Mode.CONSTANT
+
+    def __eq__(self, other):
```

Comment:
Note: I had to add this for every type, otherwise, `equals` doesn't work.

Reply:
The result for `i1 == i2` was `True`, instead of Boolean(value=True) and the tests failed because it wasn't a `NadaType`.

Reply:
We talked about this, but we are not sure why this occurs for `__eq__` only. We keep it as is right now.